src/rust/engine/src/interning.rs

```diff
@@ -51,10 +51,13 @@ impl Interns {
     Interns::default()
   }
 
-  pub fn key_insert(&self, py: Python, v: Value) -> PyResult<Key> {
+  pub fn key_insert(&self, py: Python, v: PyObject) -> PyResult<Key> {
     let (intern_key, type_id) = {
-      let obj = (*v).as_ref(py);
-      (InternKey(obj.hash()?, v.clone()), obj.get_type().into())
+      let obj = v.as_ref(py);
+      (
+        InternKey(obj.hash()?, v.clone_ref(py)),
```

Comment:
Note the `clone_ref` will increment Python's reference count, whereas before we used `.clone()` to only clone the `Arc`.